File: OCR-Translate/Services/VocabularyService.swift
import Foundation

struct VocabEntry: Equatable, Codable {
    let word: String
    let reading: String
    let meaning: String
    let partOfSpeech: String?
    let jlptLevel: String?
    let examples: [String]
    let notes: String?
}

enum VocabError: LocalizedError {
    case emptyWord
    case apiError(String)
    case parseError(String)

    var errorDescription: String? {
        switch self {
        case .emptyWord: "请输入要查询的单词"
        case .apiError(let msg): "查询出错: \(msg)"
        case .parseError(let msg): "解析出错: \(msg)"
        }
    }
}

final class VocabularyService {
    private let deepSeek = DeepSeekService()
    private let systemPrompt = """
    你是一位专业的日语词典助手。用户会输入一个日语单词或短语，请返回 JSON：

    {
      "word": "原词",
      "reading": "读音（平假名）",
      "meaning": "中文释义",
      "partOfSpeech": "词性",
      "jlptLevel": "JLPT级别",
      "examples": ["例句1（日文）", "例句2（日文）"],
      "notes": "用法说明或常见搭配"
    }

    要求：
    - reading 必须使用平假名
    - meaning 使用简洁中文
    - examples 提供 1-3 个实际例句（日文即可）
    - 如果查询的是短语，说明其构成和用法
    - 只输出 JSON
    """

    private var cache: [String: VocabEntry] = [:]

    func lookup(_ word: String) async throws -> VocabEntry {
        let trimmed = word.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { throw VocabError.emptyWord }

        let key = trimmed.lowercased()
        if let cached = cache[key] { return cached }

        let raw: String
        do {
            raw = try await deepSeek.chat(
                systemPrompt: systemPrompt,
                userMessage: trimmed,
                temperature: 0.1,
                maxTokens: 1024
            )
        } catch {
            throw VocabError.apiError(error.localizedDescription)
        }

        let entry = try parse(raw, word: trimmed)
        cache[key] = entry
        return entry
    }

    func clearCache() { cache.removeAll() }

    private func parse(_ jsonString: String, word: String) throws -> VocabEntry {
        let cleaned = jsonString
            .replacingOccurrences(of: "```json", with: "")
            .replacingOccurrences(of: "```", with: "")
            .trimmingCharacters(in: .whitespacesAndNewlines)

        guard let data = cleaned.data(using: .utf8) else {
            throw VocabError.parseError("编码失败")
        }
        guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw VocabError.parseError("JSON 解析失败")
        }

        return VocabEntry(
            word: json["word"] as? String ?? word,
            reading: json["reading"] as? String ?? "",
            meaning: json["meaning"] as? String ?? "",
            partOfSpeech: json["partOfSpeech"] as? String,
            jlptLevel: json["jlptLevel"] as? String,
            examples: json["examples"] as? [String] ?? [],
            notes: json["notes"] as? String
        )
    }
}

File: win/NGMproofread.Windows/Services/VocabularyService.cs
using System.Text.Json;
using NGMproofread.Windows.Models;

namespace NGMproofread.Windows.Services;

public class VocabEntry
{
    public string Word { get; init; } = "";
    public string Reading { get; init; } = "";
    public string Meaning { get; init; } = "";
    public string? PartOfSpeech { get; init; }
    public string? JlptLevel { get; init; }
    public List<string> Examples { get; init; } = [];
    public string? Notes { get; init; }
}

public class VocabularyService
{
    private readonly AIService _ai = new();
    private readonly Dictionary<string, VocabEntry> _cache = new();

    public async Task<VocabEntry> LookupAsync(string word)
    {
        if (_cache.TryGetValue(word, out var cached))
            return cached;

        var systemPrompt = """
            你是一个日语词典。请对用户查询的日语单词或短语给出详细解释。

            以 JSON 格式返回：
            {
              "word": "原始单词",
              "reading": "平假名读音",
              "meaning": "中文释义",
              "part_of_speech": "词性（如：名词、他动词、形容动词、副词、助词等）",
              "jlpt_level": "JLPT等级（N1-N5，不确定填 null）",
              "examples": ["例句1", "例句2"],
              "notes": "用法说明或补充信息（如无则填 null）"
            }

            只返回 JSON，不要额外解释。
            """;

        var raw = await _ai.ChatAsync(systemPrompt, word, temperature: 0.3, maxTokens: 2048);

        try
        {
            var json = raw.Trim();
            if (json.StartsWith("```")) json = json[json.IndexOf('\n')..].Trim();
            if (json.EndsWith("```")) json = json[..json.LastIndexOf("```")].Trim();

            var entry = JsonSerializer.Deserialize<VocabEntry>(json, new JsonSerializerOptions
            {
                PropertyNameCaseInsensitive = true
            }) ?? new VocabEntry { Word = word, Reading = word, Meaning = raw };

            _cache[word] = entry;
            return entry;
        }
        catch
        {
            var fallback = new VocabEntry
            {
                Word = word,
                Reading = word,
                Meaning = raw.Trim(),
                Notes = null
            };
            _cache[word] = fallback;
            return fallback;
        }
    }
}
